starky2/src/stark.rs

```diff
@@ -0,0 +1,229 @@
+    const PUBLIC_INPUTS: usize;
+
+    /// Evaluate constraints at a vector of points.
+    ///
+    /// The points are elements of a field `FE`, a degree `D2` extension of `F`. This lets us
+    /// evaluate constraints over a larger domain if desired. This can also be called with `FE = F`
+    /// and `D2 = 1`, in which case we are using the trivial extension, i.e. just evaluating
+    /// constraints over `F`.
+    fn eval_packed_generic<FE, P, const D2: usize>(
+        &self,
+        vars: StarkEvaluationVars<FE, P, { Self::COLUMNS }, { Self::PUBLIC_INPUTS }>,
+        yield_constr: &mut ConstraintConsumer<P>,
+    ) where
+        FE: FieldExtension<D2, BaseField = F>,
+        P: PackedField<Scalar = FE>;
+
+    /// Evaluate constraints at a vector of points from the base field `F`.
+    fn eval_packed_base<P: PackedField<Scalar = F>>(
+        &self,
+        vars: StarkEvaluationVars<F, P, { Self::COLUMNS }, { Self::PUBLIC_INPUTS }>,
+        yield_constr: &mut ConstraintConsumer<P>,
+    ) {
+        self.eval_packed_generic(vars, yield_constr)
+    }
+
+    /// Evaluate constraints at a single point from the degree `D` extension field.
+    fn eval_ext(
+        &self,
+        vars: StarkEvaluationVars<
+            F::Extension,
+            F::Extension,
+            { Self::COLUMNS },
+            { Self::PUBLIC_INPUTS },
+        >,
+        yield_constr: &mut ConstraintConsumer<F::Extension>,
+    ) {
+        self.eval_packed_generic(vars, yield_constr)
+    }
+
+    /// Evaluate constraints at a vector of points from the degree `D` extension field. This is like
+    /// `eval_ext`, except in the context of a recursive circuit.
+    /// Note: constraints must be added through`yeld_constr.constraint(builder, constraint)` in the
+    /// same order as they are given in `eval_packed_generic`.
+    fn eval_ext_recursively(
+        &self,
+        builder: &mut CircuitBuilder<F, D>,
+        vars: StarkEvaluationTargets<D, { Self::COLUMNS }, { Self::PUBLIC_INPUTS }>,
+        yield_constr: &mut RecursiveConstraintConsumer<F, D>,
+    );
+
+    /// The maximum constraint degree.
+    fn constraint_degree(&self) -> usize;
+
+    /// The maximum constraint degree.
+    fn quotient_degree_factor(&self) -> usize {
+        1.max(self.constraint_degree() - 1)
+    }
+
+    /// Computes the FRI instance used to prove this Stark.
+    fn fri_instance(
+        &self,
+        zeta: F::Extension,
+        g: F,
+        degree_bits: usize,
+        num_ctl_zs: usize,
+        config: &StarkConfig,
+    ) -> FriInstanceInfo<F, D> {
+        let no_blinding_oracle = FriOracleInfo { blinding: false };
+        let mut oracle_indices = 0..;
+
+        let trace_info =
+            FriPolynomialInfo::from_range(oracle_indices.next().unwrap(), 0..Self::COLUMNS);
+
+        let num_permutation_batches = self.num_permutation_batches(config);
+        let permutation_ctl_zs_info = (num_permutation_batches + num_ctl_zs > 0).then(|| {
```

Comment:
I think `num_ctl_zs > 0` should always hold? (Except maybe if this was used for single-table STARKs, but I think we can ignore other use cases since this is sort of EVM specific.)

Reply:
I guess in some designs we would have tables that aren't connected (via a lookup) to any other table. For example, an ECDSA table would be disconnected from the EVM tables (assuming the ECDSA table only verify the tx signature and that it doesn't use a lookup to e.g. a bignum table). Though in this case we could also just compute the proof as a regular `starky` proof.
What do you think? Do you think we should assume that every STARK in `AllStark` is part of a cross-table lookup?

Reply:
Hm yeah I guess we could use starky if we did have an "isolated" table. Also I think we'll probably want to connect ECDSA to the CPU table, to support ecrecover and maybe a variable number of txns in one proof.

Reply:
Makes sense, so do you think we should assume `num_ctl_zs > 0` or keep the current level of generality?

Reply:
Hm it seems like assuming it could lead to several small simplifications (like removing a few conditions and unwraps in prover), so I think it's probably worth doing unless/until a use case comes up where we wanted an isolated table (I can't think of one right now).

Reply:
Sounds good, I'll make the change.